add hub.config passthrough and use it for all auth config (#1943)
jupyterhub/zero-to-jupyterhub-k8s: doc/source/administrator/authentication.md

(authentication)=

# Authentication and authorization

Authentication is about identity, while _authorization_ is about permissions. In
this section you will learn how to configure both. As an example, you can
configure authentication using GitHub accounts and restrict what users are
authorized based on membership of a GitHub organization.

Before configuring this, you should have [setup HTTPS](https).

## Useful understanding

### Authenticator classes

> JupyterHub are by itself unable to authenticate any kind of identity, but
> instead relies on _one_ chosen [_authenticator
> class_](https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html)
> to do it. Several such classes are already made available through [installed

Reviewer: ```suggestion
JupyterHub by default ships with only one source of authentication:
PAM, the underlying unix authentication of the host system.
To use other sources of authentication, choose _one_ [_authenticator
class_](https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html) to use.
Several such classes are already available in the hub image through [installed
```

Python
packages](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/images/hub/requirements.txt).

JupyterHub provides a base class,
[`Authenticator`](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/auth.py),
that all other authenticator classes are supposed to derive from. By configuring
this base class, we influence the behavior of the derived class as well.

### The configuration system

We configure JupyterHub to use our chosen authenticator class and the
authenticator class itself through this Helm chart's
[`hub.config`](schema_hub.config) configuration.

## General configuration

> As all authenticators classes derive from the `Authenticator` base class, they

Reviewer: ```suggestion
As all authenticator classes derive from the `Authenticator` base class, they
```

share some configuration options. Below are some common configuration options,
but please refer to the official [configuration
reference](https://jupyterhub.readthedocs.io/en/latest/api/auth.html) for more
details.

### [allowed_users](https://jupyterhub.readthedocs.io/en/latest/api/auth.html#jupyterhub.auth.Authenticator.allowed_users) / [admin_users](https://jupyterhub.readthedocs.io/en/latest/api/auth.html#jupyterhub.auth.LocalAuthenticator.admin_users)

Some authenticator classes may have dedicated logic in addition this this to
authorize users.

```yaml
hub:
  config:
    Authenticator:
      admin_users:
        - user1
        - user2
      allowed_users:
        - user3
        - user4
    # ...
    DummyAuthenticator:
      password: a-shared-secret-password
    JupyterHub:
      authenticator_class: dummyauthenticator.DummyAuthenticator
```

> 

Reviewer: We can add some explanation to these configuration examples:

```suggestion

In the above configuration, we have configured three things:

1. JupyterHub is instructed to use the dummy authenticator to login (only appropriate for testing purposes),
2. anyone will be able to login with username `user1-4` and the password `a-shared-secret-password`
3. `user1` and `user2` will have admin permissions, while `user3` and `user4` will be regular users.

```

### [auto_login](https://jupyterhub.readthedocs.io/en/latest/api/auth.html#jupyterhub.auth.Authenticator.auto_login)

If you have configured authentication with GitHub for example, the page
`/hub/login` will feature a single orange button that users are to press to

> login. If you want to bypass this screen and redirect users directly, you can

Reviewer: ```suggestion
login. If you want to bypass this screen and send users directly to GitHub login, you can
```

_identity provider_ such as GitHub, Google, and CILogon. All of these will
require an OAuth2 _client id_ and _client secret_.

For details on how to acquire a client id and client secret, please refer to
[oauthenticator's
documentation](https://oauthenticator.readthedocs.io/en/stable/getting-started.html).

#### GitHub

GitHub is the largest hosting service for git repositories. It is free to create
an account at GitHub, and relatively straightforward to set up OAuth credentials
so that users can authenticate with their GitHub username/password.

To create OAuth credentials on GitHub, follow these steps:

-   Click your profile picture -> settings -> developer settings
-   Make sure you're on the "OAuth Apps" tab, then click "New OAuth App"
-   Fill out the forms (you'll need your hub address) and generate your
    ID/Secret.

To enable GitHub authentication, your `config.yaml` should contain the following
configuration:

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
    JupyterHub:

>       authenticator_class: oauthenticator.github.GitHubOAuthenticator

Reviewer: ```suggestion
      authenticator_class: github
```

```

Make sure that the `oauth_callback_url` matches the one you set in GitHub.

To restrict access to the members of one or more GitHub organizations, amend
your previous configuration with these parts.

```yaml
hub:
  config:
    GitHubOAuthenticator:
      allowed_organizations:
        - my-github-organization

>     OAuthenticator:

Reviewer: While we *can* set every option on the highest possible class, it is not necessary. It's fine and probably easier to follow to only directly configure the class being used in many of these cases:

```suggestion
```

    type" > "Web application".
5.  Enter a name for your JupyterHub instance. You can give it a descriptive
    name or set it to be the hub's hostname.
6.  Set "Authorized JavaScript origins" to be your hub's URL.
7.  Set "Authorized redirect URIs" to be your hub's URL followed by
    `/hub/oauth_callback`. For example,
    `https://your-jupyterhub-domain/hub/oauth_callback`.
8.  When you click "Create", the console will generate and display a Client ID
    and Client Secret. Save these values.
9.  Type "consent screen" in the search field at the top and click to access the
    OAuth consent screen. Here you will customize what your users see when they
    login to your JupyterHub instance for the first time. Click Save when you
    are done.
10. Update your Helm chart's configuration (`config.yaml`) to look like this.

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id.apps.googleusercontent.com
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
    GoogleOAuthenticator:
      hosted_domain:
        - your-university.edu
      login_service: Your university
    JupyterHub:

>       authenticator_class: oauthenticator.google.GoogleOAuthenticator

Reviewer: ```suggestion
      authenticator_class: google
```

```

The `oauth_callback_url` key is set to the authorized redirect URI you specified
earlier. Set `hosted_domain` to your institution's domain name. The value of
`login_service` is a descriptive term for your institution that reminds your
users which account they are using to login.

#### CILogon

Please see CyberInfrastructure Logon's [website](https://www.cilogon.org) for
more information about what kind of identity is managed by CILogon.

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
    JupyterHub:

>       authenticator_class: oauthenticator.cilogon.CILogonOAuthenticator

Reviewer: ```suggestion
      authenticator_class: cilogon
```

>   CILogonOAuthenticator:
>     CiOAuthenticator:

Reviewer: typo:

```suggestion
 config:
    CILogonOAuthenticator:
```

(authentication)=

# Authentication and authorization

Authentication is about identity, while _authorization_ is about access rights.
In this section you will learn how to configure both. For example, you can
configure to authenticate GitHub users and authorize those part of a specific

> GitHub organization.

Reviewer 2: Not sure, but I think you're trying to say
"You can configure authentication using GitHub accounts, but restrict what users are authorised to do (or what users can log in?) based on membership of a GitHub organiisation"
?

Author: What do you think about 3ea9ded0?

Before configuring this, you should have [setup HTTPS](https).

## Useful understanding

### Authenticator classes

As JupyterHub can't know how to authenticate all kinds of users, it expose
configuration to extend its capabilities to rely on a dedicated [_authenticator
class_](https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html).
Several such classes are already made available through [installed Python

> packages](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/images/hub/requirements.txt).

Reviewer 2: Is there any magic in MyST that would allow the tag to appear here instead of `master`? So that the stable docs corresponding to a release could link to the correct set of packages, whereas the dev/master docs could link to `master`?

Author: This is now resolved through https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1968/files#diff-d8d3ed25802824d15bf411f8e97416d8fc6f9247e821b0617f2b869dc584b99cR102-R109.

### The Authenticator base class

JupyterHub provides a base class,
[`Authenticator`](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/auth.py),
that all other authenticator classes are supposed to derive from. By configuring
this base class, we influence the behavior of the derived class as well.

### The configuration system

> We configure the JupyterHub to use our chosen authenticator class and the

Reviewer 2: ```suggestion
We configure JupyterHub to use our chosen authenticator class and the
```
Or maybe "the JupyterHub deployment"?

authenticator class in question through this Helm chart's
[`hub.config`](schema_hub.config) configuration.

## Configuring authorization

Some authenticator classes contain dedicated authorization logic, but that
doesn't stop you from using the common base class authorization logic.

```yaml
hub:
  config:
    Authenticator:
      admin_users:
        - user1
        - user2
      allowed_users:
        - user3
        - user4
    # ...
    DummyAuthenticator:
      password: a-shared-secret-password
    JupyterHub:
      authenticator_class: dummyauthenticator.DummyAuthenticator
```

```{note}
Please refer to [JupyterHub's own documentation](https://jupyterhub.readthedocs.io/en/latest/getting-started/authenticators-users-basics.html) for more details.
```

## Configuring authenticator classes

### OAuth2 based authentication

JupyterHub's
[oauthenticator](https://github.com/jupyterhub/oauthenticator) has
support for enabling your users to authenticate via a third-party OAuth
provider, including GitHub, Google, and CILogon.

Follow the service-specific instructions linked on the [oauthenticator

> repository](https://github.com/jupyterhub/oauthenticator) to generate

Reviewer 2: There's also a readthedocs site https://oauthenticator.readthedocs.io/en/stable/ , does this also have the required information?

Author: ```suggestion
repository](https://oauthenticator.readthedocs.io/en/stable/) to generate
```

your JupyterHub instance's OAuth2 client ID and client secret. Then
declare the values in the helm chart (`config.yaml`).

Here are example configurations for common authentication services. Note

> that in each case, you need to get the authentication credential
> information before you can configure the helm chart for authentication.

Reviewer 2: ```suggestion
that in each case, you need to get the client ID and secret
before you can configure the helm chart for authentication.
```
I think repeating this is clearer than introducing a new term "authentication credential information"

Public](https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/publicizing-or-hiding-organization-membership)
to login, whereas setting it to `read:org` may require approval of the
application by a GitHub organisation admin. Please see [this
issue](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/687)
for further information.

```{note}
Changing `scope` will not change the scope for existing OAuth tokens,
you must invalidate them.
```

#### Google

Google authentication is used by many universities (it is part of the
"G Suite"). Note that using Google authentication requires your Hub to
have a domain name (it cannot **only** be accessible via an IP address).

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id.apps.googleusercontent.com
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
    GoogleOAuthenticator:
      hosted_domain:
        - your-university.edu
      login_service: Your university
    JupyterHub:
      authenticator_class: oauthenticator.GoogleOAuthenticator
```

#### CILogon

> 

Reviewer 2: Could you add a couple of sentences on what CILogon is, since it's the only probvider that doesn't have any background.

Author: ```suggestion
Please see CyberInfrastructure Logon's [website](https://www.cilogon.org) for more information about what kind of identity is managed by CILogon.

```

login button.

Here's an example for authenticating against
[keycloak](https://www.keycloak.org/docs/latest/securing_apps/index.html#endpoints),
after you [configure an OIDC
Client](https://www.keycloak.org/docs/latest/server_admin/index.html#oidc-clients)
and obtain the confidential client credentials.

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
      authorize_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/auth
      token_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/token
      userdata_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/userinfo
    GenericOAuthenticator:
      login_service: keycloak
      username_key: preferred_username
      userdata_params:
        state: state
    JupyterHub:
      authenticator_class: oauthenticator.generic.GenericOAuthenticator
```

#### Auth0

Auth0 is a popular commercial provider of identity management. The
JupyterHub helm chart does not include support for Auth0 by default. To

> use Auth0, `extraEnv` and `extraConfig` must be configured as follows:

Reviewer 2: You've mentioned `extraEnv` and `extraConfig` but there's no sign of it below, maybe leftover from a copy and paste?

Author: Ah, I removed extraEnv and perhaps also extraConfig without updating the text. I'll update the text to match.

> Note that without the scope defined, authenticating to JupyterHub after

Reviewer 2: ```suggestion
Note that without the scopes defined, authenticating to JupyterHub after
```

authentication.

1.  Log in to the [Google API
    Console](https://console.developers.google.com).
2.  Select a project > Create a project\... and set 'Project name'.
    This is a short term that is only displayed in the console. If you
    have already created a project you may skip this step.
3.  Type "Credentials" in the search field at the top and click to
    access the Credentials API.
4.  Click "Create credentials", then "OAuth client ID". Choose
    "Application type" > "Web application".
5.  Enter a name for your JupyterHub instance. You can give it a
    descriptive name or set it to be the hub's hostname.
6.  Set "Authorized JavaScript origins" to be your hub's URL.
7.  Set "Authorized redirect URIs" to be your hub's URL followed by
    `/hub/oauth_callback`. For example,
    `https://your-jupyterhub-domain/hub/oauth_callback`.
8.  When you click "Create", the console will generate and display a
    Client ID and Client Secret. Save these values.
9.  Type "consent screen" in the search field at the top and click to
    access the OAuth consent screen. Here you will customize what your
    users see when they login to your JupyterHub instance for the first

>     time. Click Save when you are done.

Reviewer 2: As an aside, the above steps would be really useful for non Z2JH users too! How do you feel about linking to this doc from other project docs?

Author: Sure why not. One answer would be to avoid locking in whats written exactly here by letting it be referenced by other projects.

I think i'd like to see some refactoring of this page in general, and preferably before referencing it since the structure may change a bit based on the refactoring.

As an example of refactoring, I think it could be nice to have a common section of the OAuthenticator configs, or link out to oauthenticator where we improve the docs of that etc. Not sure!

Author: I did some refactoring. I think in general, it would be good if we let OAuthenticator maintain the docs on how to setup client id etc, as it is used by z2jh / tljh and vanilla JupyterHub's etc, while z2jh is just one distribution where its relevant.

10. In your helm chart, create a stanza that contains these OAuth
    fields:

```yaml
hub:
  config:
    OAuthenticator:
      client_id: your-client-id.apps.googleusercontent.com
      client_secret: your-client-secret
      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
    GoogleOAuthenticator:
      hosted_domain:
        - your-university.edu
      login_service: Your university
    JupyterHub:

>       authenticator_class: oauthenticator.GoogleOAuthenticator

Reviewer 2: In other places you've used `oauthenticator.MODULE.XXXAuthenticator` instead of `oauthenticator.XXXAuthenticator`. Either works, but I think it'd be good to be consistent. Alternatively can an entrypoint be specified?

Author: @manics oh, I have just copied the jupyterhub_config.py files mapping. I assume this relates to the __init__.py file of the oauthenticator modules folder? I suggest we keep it short but not using `google` or similar as it hides the fact its related to the oauthenticator python package which I want us to highlight.

Author: https://github.com/jupyterhub/oauthenticator/blob/6f239bebecbb3fb0242de7f753ae1c93ed101340/oauthenticator/__init__.py#L1-L7

It seems that only a few are added there for backward compatebility. So, we can only make all longer. Let's make all longer.

Author: Done.

      authenticator_class: ldapauthenticator.LDAPAuthenticator
    LDAPAuthenticator:
      bind_dn_template:
        - cn={username},ou=edir,ou=people,ou=EXAMPLE-UNIT,o=EXAMPLE
      server_address: ldap.EXAMPLE.org
```

#### Example Active Directory Configuration

This example is equivalent to that given in the [ldapauthenticator
README](https://github.com/jupyterhub/ldapauthenticator/blob/master/README.md).

```yaml
hub:
  config:
    JupyterHub:
      authenticator_class: ldapauthenticator.LDAPAuthenticator
    LDAPAuthenticator:
      allowed_groups:
        - cn=researcher,ou=groups,dc=wikimedia,dc=org
        - cn=operations,ou=groups,dc=wikimedia,dc=org
      bind_dn_template:
        - uid={username},ou=people,dc=wikimedia,dc=org
        - uid={username},ou=developers,dc=wikimedia,dc=org
      escape_userdn: false
      lookup_dn: true
      lookup_dn_search_filter: ({login_attr}={login})
      lookup_dn_search_password: secret
      lookup_dn_search_user: ldap_search_user_technical_account
      lookup_dn_user_dn_attribute: cn
      server_address: ad.EXAMPLE.org
      user_attribute: sAMAccountName
      user_search_base: ou=people,dc=wikimedia,dc=org
```

> #### Example Auth0 Configuration

Reviewer 2: I'm a bit confused, there's an earlier ` #### Auth0` section that uses `oauthenticator.auth0.Auth0OAuthenticator` but this is different? Is to illustrate OpenID (and Auth0 just happens to be an example OpenID provider)?

Author: I'm also confused, I decided to not care about this difference as part of this PR.

Author: I went ahead and updated the entire authentication page where I located two examples under a GenericOAuthenticator section, where I took Auth0 and KeyCloak examples against OpenID connect.
